### PR TITLE
Access package control

### PR DESCRIPTION
**Pull request by OlaStalberg**

### CODE REVIEW

File: src/data/api/EntityApi.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class EntityApi {
+
+    static getEntities() {
+        const url = "/api/components/configurations";
+
+        return fetch(url, {
+            method: 'GET',
+            credentials: 'same-origin',
+            headers: {
```
Comment (fsjovatsen) [inline src/data/api/EntityApi.js:9]: Think this can be removed

File: src/data/redux/actions/entity.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import {ADD_COMPONENT_CONFIGURATION, GET_ENTITIES_ERROR} from "./types";
+
+
+export function fetchEntitiesSuccess(jsonData) {
+    return {
+        type: ADD_COMPONENT_CONFIGURATION,
```
Comment (fsjovatsen) [inline src/data/redux/actions/entity.js:6]: Missing symmetry in actions.

File: src/data/redux/actions/entity.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+        payload: jsonData
+    }
+}
+
+
+export function fetchEntitiesError() {
+    return {
+        type: GET_ENTITIES_ERROR,
```
Comment (fsjovatsen) [inline src/data/redux/actions/entity.js:14]: Missing symmetry in actions.

File: src/data/redux/reducers/component-configuration.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import {ADD_COMPONENT_CONFIGURATION} from "../actions/types";
+
+export default function component_configuration(state = [], action) {
+    switch (action.type) {
+        case ADD_COMPONENT_CONFIGURATION:
```
Comment (fsjovatsen) [inline src/data/redux/reducers/component-configuration.js:5]: Missing symmetry in actions.

File: src/features/access-package/access_package_container.jsx [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+    useEffect(() => {
+            dispatch(fetchAccess(context.currentOrganisation.name));
+            dispatch(fetchComponents());
+            dispatch(fetchEntities());
+            dispatch(fetchClients(context.currentOrganisation.name));
+        }, [dispatch, context.currentOrganisation.name]
+    );
+
+    if (!componentConfiguration || componentConfiguration.length === 0 || !access) {
```
Comment (fsjovatsen) [inline src/features/access-package/access_package_container.jsx:26]: Maybe put the if expression in a method to document it?

File: src/features/access-package/access_package_list.jsx [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    const [openSave, setOpenSave] = useState(false);
+    const [openDialog, setOpenDialog] = React.useState(false);
+    const [packageToDelete, setPackageToDelete] = React.useState(null);
+    const [snackBarOpen, setSnackBarOpen] = useState(false);
+    const [snackBarMessage, setSnackBarMessage] = React.useState("");
+    const context = useContext(AppContext);
+    const _ = require('lodash');
+    const deleteMessageText = packageToDelete != null ? "Vil du fjerne tilgangspakken " + packageToDelete.name + "?" : "";
+
+    let selectedAccessPackage = undefined;
```
Comment (fsjovatsen) [inline src/features/access-package/access_package_list.jsx:68]: should this be a state?

File: src/features/access-package/access_package_list.jsx [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    packages.map(ap => {
+        if (ap.dn === selectedForEditingId) {
```
Comment (fsjovatsen) [inline src/features/access-package/access_package_list.jsx:70]: use filter

File: src/features/access-package/add/add_access_package_form.jsx [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+    function updateDescription(event) {
+        setDescription(event.target.value);
+    }
+
+    function handleCreatePackage() {
+        const access = {};
+        access.name = name;
+        access.description = description;
+        AccessApi.setAccess(access, context.currentOrganisation.name)
```
Comment (fsjovatsen) [inline src/features/access-package/add/add_access_package_form.jsx:30]: maybe set the object right in? It's small.

File: src/features/access-package/edit/changed_accesses.jsx [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                <ListItemText
+                    primary={entity}
+                />
+            </ListItem>
+        );
+    }
+
+    return (
+        <div>
+            <List dense>
```
Comment (fsjovatsen) [inline src/features/access-package/edit/changed_accesses.jsx:57]: break this up in methods to make it more readable.

File: src/features/access-package/edit/changed_accesses.jsx [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    const {oldAccessPackage, newAccessPackage} = props;
+    const componentConfiguration = useSelector(state => state.component_configuration.componentConfiguration);
+    const collectionList = [];
+    const readList = [];
+    const modifyList = [];
+
+    function accessListItem(entry) {
+        const removingEntry = entry.charAt(0) === "-";
+        let entity = entry.substring(2, entry.length);
+        componentConfiguration.map(componentConfiguration => {
```
Comment (fsjovatsen) [inline src/features/access-package/edit/changed_accesses.jsx:33]: extract to method?

File: src/features/access-package/edit/changed_clients.jsx [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+}));
+
+const ChangedClients = (props) => {
+    const {oldAccessPackage, newAccessPackage} = props;
+    const classes = useStyles();
+    const clientList = [];
+    return (
+        <div>
+            <List dense>
+                {newAccessPackage.clients.forEach(entry => {
```
Comment (fsjovatsen) [inline src/features/access-package/edit/changed_clients.jsx:27]: extract to methods?

File: src/features/access-package/edit/changed_components.jsx [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+const ChangedComponents = (props) => {
+    const classes = useStyles();
+    const {oldAccessPackage, newAccessPackage} = props;
+    const componentConfiguration = useSelector(state => state.component_configuration.componentConfiguration);
+    const componentList = [];
+
+    function accessListItem(entry) {
+        const removingEntry = entry.charAt(0) === "-";
+        let entity = entry.substring(2, entry.length);
+        componentConfiguration.map(componentConfiguration => {
```
Comment (fsjovatsen) [inline src/features/access-package/edit/changed_components.jsx:31]: extract to methods?

File: src/features/access-package/edit/changed_components.jsx [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+                    primary={entity}
+                />
+            </ListItem>
+        );
+    }
+
+    return (
+        <div>
+            <List dense>
+                {newAccessPackage.components.forEach(entry => {
```
Comment (fsjovatsen) [inline src/features/access-package/edit/changed_components.jsx:53]: extract to methods?

File: src/features/access-package/edit/client_selection.jsx [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+    const accessPackages = useSelector(state => state.access_package.accessPackages);
+    const [showWarning, setShowWarning] = useState(false);
+    const [switchValue, setSwitchValue] = useState(false);
+    const [switchClient, setSwitchClient] = useState(null);
+    const warningMessageText = "Når du aktiverer en aksesspakke på en klient, vil andre aksesspakker som er koblet til denne klienten bli fjernet. Ønsker du å fortsette?";
+
+
+    const dispatch = useDispatch();
+
+    function findIndex(array, value) {
```
Comment (fsjovatsen) [inline src/features/access-package/edit/client_selection.jsx:55]: use `filter`

File: src/features/access-package/edit/client_selection.jsx [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+        let newAccessPackage = {...selectedAccessPackage};
+        const accessPackageIndex = findIndex(newAccessPackages, newAccessPackage);
+        if (newClients.includes(client.dn)) {
+            let clientIndex = newClients.indexOf(client.dn);
+            newClients.splice(clientIndex, 1);
+        } else {
+            newClients.push(client.dn);
+        }
+        newAccessPackage.clients = newClients;
+        newAccessPackages[accessPackageIndex] = newAccessPackage;
```
Comment (fsjovatsen) [inline src/features/access-package/edit/client_selection.jsx:83]: can possibly be -1 and give an out of index...

File: src/features/access-package/edit/edit_access_package_container.jsx [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+        if (newAccessPackage.components.includes(component.dn)) {
+            let componentIndex = findComponentIndex(newAccessPackage.components, component.dn);
+            newAccessPackage.components.splice(componentIndex, 1);
+            newAccessPackage.collection = removePathsFromList(newAccessPackage.collection, component);
+            newAccessPackage.read = removePathsFromList(newAccessPackage.read, component);
+            newAccessPackage.modify = removePathsFromList(newAccessPackage.modify, component);
+        } else {
+            newAccessPackage.components.push(component.dn);
+        }
+        newAccessPackages[accessPackageIndex] = newAccessPackage;
```
Comment (fsjovatsen) [inline src/features/access-package/edit/edit_access_package_container.jsx:114]: may be -1

File: src/features/access-package/view/add_access_package_to_client.jsx [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+                                />
+                            ))
+                            }
+                        </List>
+                    </div>
+                </div>
+            </div>
+        );
+    } else {
+        return (<></>);
```
Comment (fsjovatsen) [inline src/features/access-package/view/add_access_package_to_client.jsx:117]: should be `</>`

**Review by fsjovatsen — CHANGES_REQUESTED** [review]

File: src/data/api/EntityApi.js (hunk 1 above)
Comment (OlaStalberg) [reply src/data/api/EntityApi.js:9]: Removed

File: src/data/redux/actions/entity.js (hunk 2 above)
Comment (OlaStalberg) [reply src/data/redux/actions/entity.js:6]: Resolved

File: src/data/redux/actions/entity.js (hunk 3 above)
Comment (OlaStalberg) [reply src/data/redux/actions/entity.js:14]: Resolved

File: src/data/redux/reducers/component-configuration.js (hunk 4 above)
Comment (OlaStalberg) [reply src/data/redux/reducers/component-configuration.js:5]: Resolved

File: src/features/access-package/access_package_container.jsx (hunk 5 above)
Comment (OlaStalberg) [reply src/features/access-package/access_package_container.jsx:26]: Expression put in method

File: src/features/access-package/edit/changed_components.jsx (hunk 13 above)
Comment (OlaStalberg) [reply src/features/access-package/edit/changed_components.jsx:53]: Extracted to method

File: src/features/access-package/edit/changed_accesses.jsx (hunk 9 above)
Comment (OlaStalberg) [reply src/features/access-package/edit/changed_accesses.jsx:57]: Extracted to method

File: src/features/access-package/edit/changed_accesses.jsx (hunk 10 above)
Comment (OlaStalberg) [reply src/features/access-package/edit/changed_accesses.jsx:33]: extracted

File: src/features/access-package/edit/changed_clients.jsx (hunk 11 above)
Comment (OlaStalberg) [reply src/features/access-package/edit/changed_clients.jsx:27]: extraced

File: src/features/access-package/edit/changed_components.jsx (hunk 12 above)
Comment (OlaStalberg) [reply src/features/access-package/edit/changed_components.jsx:31]: extracted

File: src/features/access-package/edit/client_selection.jsx (hunk 14 above)
Comment (OlaStalberg) [reply src/features/access-package/edit/client_selection.jsx:55]: Removed the function replaced with filter and indexOf

File: src/features/access-package/edit/client_selection.jsx (hunk 15 above)
Comment (OlaStalberg) [reply src/features/access-package/edit/client_selection.jsx:83]: Fixed by using array indexOf and filters. removed method

File: src/features/access-package/edit/edit_access_package_container.jsx (hunk 16 above)
Comment (OlaStalberg) [reply src/features/access-package/edit/edit_access_package_container.jsx:114]: Fixed by using array indexOf and filters. removed method

File: src/features/access-package/view/add_access_package_to_client.jsx (hunk 17 above)
Comment (OlaStalberg) [reply src/features/access-package/view/add_access_package_to_client.jsx:117]: I am not allowed to do that.

**Review by fsjovatsen — APPROVED** [review]

**Pull request merged**